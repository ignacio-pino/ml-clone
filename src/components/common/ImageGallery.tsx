import {
  VStack,
  Image,
  Button,
  Flex,
  Heading,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalHeader,
  ModalOverlay,
  Divider,
  Text,
} from "@chakra-ui/react";
import { useState } from "react";
import { IPictureList } from "../../Interfaces";

export default function ImageGallery({
  pictures,
}: {
  pictures: IPictureList[];
}) {
  const [selectedImage, setSelectedImage] = useState({ id: "", url: "" });
  const [isModalOpen, setIsModalOpen] = useState(false);

  let shortenedPictures: IPictureList[] = [];

  if (pictures.length > 6) {
    shortenedPictures = pictures.slice(0, 6);
  }

  return (
    <>
      <Flex>
        <VStack marginRight={"5%"}>
          {(shortenedPictures.length > 0 ? shortenedPictures : pictures).map(
            ({ id, url }) => (
              <Button
                key={id}
                onClick={() => setSelectedImage({ id, url })}
                w={"fit-content"}
                h={"fit-content"}
                p={1}
                variant={"outline"}
                boxShadow={"2xl"}
                data-testid={`imageButton-${url}`}
              >
                <Image
                  src={url}
                  h={"50px"}
                  w={"50px"}
                  alt={`imageButton-${id}`}
                />
              </Button>
            )
          )}
          {shortenedPictures.length > 0 && (
            <Button
              onClick={() => setIsModalOpen(true)}
              w={"50px"}
              h={"50px"}
              p={1}
              variant={"outline"}
              boxShadow={"2xl"}
              data-testid="showMoreImagesButton"
            >
              <Text fontSize={"lg"}>+{pictures.length - 6} </Text>
            </Button>
          )}
        </VStack>
        <Flex justifyContent={"center"} w={"500px"}>
          <Image
            borderRadius={"5px"}
            src={selectedImage.url ? selectedImage.url : pictures[0].url}
            key={selectedImage.id ? selectedImage.id : pictures[0].id}
            boxShadow={"dark-lg"}
            maxW={"500px"}
            maxH={"500px"}
            alt={selectedImage.id ? selectedImage.id : pictures[0].id}
          />
        </Flex>
      </Flex>
      <Modal
        isOpen={isModalOpen}
        onClose={() => setIsModalOpen(false)}
        size={"xl"}
        scrollBehavior="inside"
      >
        <ModalOverlay />
        <ModalContent>
          <ModalHeader>
            <Heading>Im√°genes del producto</Heading>
            <ModalCloseButton />
          </ModalHeader>
          <ModalBody>
            {pictures.map(({ id, url }) => (
              <VStack data-testid="imageList">
                <Image
                  key={id}
                  w={"fit-content"}
                  h={"fit-content"}
                  p={1}
                  boxShadow={"2xl"}
                  src={url}
                  alt={id}
                />
                <Divider my={"2rem"} backgroundColor={"gray.500"} />
              </VStack>
            ))}
          </ModalBody>
        </ModalContent>
      </Modal>
    </>
  );
}
