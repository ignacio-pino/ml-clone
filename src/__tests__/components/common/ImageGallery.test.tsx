import { render, screen, fireEvent } from "@testing-library/react";
import ImageGallery from "../../../components/common/ImageGallery";

const samplePictures = [
  { id: "1", url: "url1" },
  { id: "2", url: "url2" },
  { id: "3", url: "url3" },
  { id: "4", url: "url4" },
  { id: "5", url: "url5" },
  { id: "6", url: "url6" },
];

describe("ImageGallery Component", () => {
  it("should match snapshot", () => {
    const container = render(
      <ImageGallery pictures={samplePictures} />
    ).container;
    expect(container).toMatchSnapshot();
  });
  it("displays main image initially as the first picture", async () => {
    render(<ImageGallery pictures={samplePictures} />);
    const mainImage = screen.getByRole("img", { name: "1" });

    expect(mainImage).toBeInTheDocument();
  });

  it("updates main image when an image button is clicked", async () => {
    render(<ImageGallery pictures={samplePictures} />);
    fireEvent.click(screen.getByTestId("imageButton-url2"));
    const mainImage = await screen.findByRole("img", { name: "2" });
    expect(mainImage).toBeInTheDocument();
  });

  it('shows a button with "+n more" when there are more than 6 images', () => {
    render(
      <ImageGallery
        pictures={[
          ...samplePictures,
          { id: "7", url: "url7" },
          { id: "8", url: "url8" },
        ]}
      />
    );
    const showMoreButton = screen.getByTestId("showMoreImagesButton");
    expect(showMoreButton).toBeInTheDocument();
    expect(showMoreButton).toHaveTextContent("+2");
  });

  it('opens a modal with all images when "+n more" button is clicked', () => {
    render(
      <ImageGallery
        pictures={[
          ...samplePictures,
          { id: "7", url: "url7" },
          { id: "8", url: "url8" },
        ]}
      />
    );
    const showMoreButton = screen.getByTestId("showMoreImagesButton");
    fireEvent.click(showMoreButton);

    samplePictures.forEach((picture) => {
      const imageInModal = screen.getByRole("img", {
        name: picture.id,
      });
      expect(imageInModal).toBeInTheDocument();
    });
  });
});
