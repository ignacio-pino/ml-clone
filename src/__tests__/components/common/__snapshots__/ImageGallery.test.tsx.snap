// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImageGallery Component > should match snapshot 1`] = `
<div>
  <div
    class="css-k008qs"
  >
    <div
      class="chakra-stack css-jlrmky"
    >
      <button
        class="chakra-button css-bilopy"
        data-testid="imageButton-url1"
        type="button"
      >
        <img
          alt="imageButton-1"
          class="chakra-image css-7f50zt"
          src="url1"
        />
      </button>
      <button
        class="chakra-button css-bilopy"
        data-testid="imageButton-url2"
        type="button"
      >
        <img
          alt="imageButton-2"
          class="chakra-image css-7f50zt"
          src="url2"
        />
      </button>
      <button
        class="chakra-button css-bilopy"
        data-testid="imageButton-url3"
        type="button"
      >
        <img
          alt="imageButton-3"
          class="chakra-image css-7f50zt"
          src="url3"
        />
      </button>
      <button
        class="chakra-button css-bilopy"
        data-testid="imageButton-url4"
        type="button"
      >
        <img
          alt="imageButton-4"
          class="chakra-image css-7f50zt"
          src="url4"
        />
      </button>
      <button
        class="chakra-button css-bilopy"
        data-testid="imageButton-url5"
        type="button"
      >
        <img
          alt="imageButton-5"
          class="chakra-image css-7f50zt"
          src="url5"
        />
      </button>
      <button
        class="chakra-button css-bilopy"
        data-testid="imageButton-url6"
        type="button"
      >
        <img
          alt="imageButton-6"
          class="chakra-image css-7f50zt"
          src="url6"
        />
      </button>
    </div>
    <div
      class="css-wxs6zv"
    >
      <img
        alt="1"
        class="chakra-image css-zjazeo"
        src="url1"
      />
    </div>
  </div>
</div>
`;
